### ISSUE TITLE
(dev2.0) Object constructor and member overloads

### ISSUE THREAD

**Issue by ncave** [fable-compiler/Fable#1472]
Currently it's a bit hard to implement BCL objects with overloading constructors/members in Fable.Core, since there is no way to know during replacement what the mangled name will be (as that happens after).
(F# modules/functions are fine as there are no overloads or interfaces to worry about.)

Would it be easier if the overloads get disambiguated (numbered with a predictable suffix) *before* they get to replacement, and then the suffix can be stripped during replacement matching, but is still there for generating the right call to the correct overload?

Just thinking aloud, please ignore if it does not make sense.

**Comment by alfonsogarciacaro** [comment]
You're right. I've never liked the way to suffix overloads that requires traversing all the members every time and is easily breakable (as with the problem we had with different members in signature files). I was also considering ways to create a predictable suffix for overloads, and I'm definitely open to do it if we find a good solution.

Funscript just attached the types of the arguments to the method name. It's an option but I didn't like it very much because the member names would become much longer. It's not a big problem because we are already mangling member names in Fable 2, and this shouldn't affect minified code. But still can make the results of the REPL somewhat less nice to see.

I'm not an expert, but do you know a way to create a unique and predictable (and not too long) hash out of the type arguments that we could attach to the member name?

**Comment by alfonsogarciacaro** [comment]
What [about this](https://github.com/fable-compiler/Fable/commit/3c6afbcec0af0202985d7ebe7518c30b2455fef1) (branch `dev2.0-overload-suffix`)? It creates an overload suffix combining the string hashes of the argument type names. Some tests still need to be fixed (those that call F# Map or Set), but do you think it could work?

**Comment by ncave** [comment]
Sure, creating a name from the parameters could work too. My suggestion above was to add a ReplaceCallInfo.MagledName for the overloads (or have a way to compute it), simple enumeration would do (e.g. Member$1, Member$2, etc.), the point is to have it available during replacement so the full mangled name can be composed, but your solution is pretty much the same thing.

**Comment by alfonsogarciacaro** [comment]
Nah, using the hash for the overloads in fable-core F# types doesn't work, because the replaced types (`Microsoft.FSharp.Collections.FSharpSet` for example) have a different names than the types in fable-core (`Set.FSharpSet`).

**Comment by alfonsogarciacaro** [comment]
The issue is it affects to the argument types as well (e.g. Set.IsProperSubsetOf also accepts another set as argument). Right now I'm considering using the hash for user-defined members and a simple enumeration for F# types in fable-core. I like the hash because it makes the name predictable, prevents the issue with signature and implementation files, doesn't require traversing the type every time and doesn't change if other members do. But maybe if introduces too much complication to have two ways to calculate the overload suffix?
